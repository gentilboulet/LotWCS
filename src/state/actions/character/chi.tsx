import { createAction } from "typesafe-actions";
import { historyMetaCreator } from "../meta";
import { ICost } from "../../models/character/costs";
import { TChiName } from "../../../data/chi";

export const chiBuy = createAction(
  "chi/BUY",
  (chi: TChiName, value: number, cost: ICost) => {
    return { chi, value, cost };
  },
  historyMetaCreator,
)();
