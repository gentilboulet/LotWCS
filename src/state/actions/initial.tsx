import * as constants from 'constants/initial';

export interface IInitialState {
  type: constants.INITIAL_STATE;
}

export type IInitialStateAction = IInitialState;

export function resetToInitialState(): IInitialState {
  return {type: constants.INITIAL_STATE};
}
