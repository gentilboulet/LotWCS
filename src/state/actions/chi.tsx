import * as constants from 'state/constants/chi';
import * as dataChi from 'data/chi';
import { ICost } from 'costs/types';

export interface IChiBuy {
  type: constants.CHI_BUY;
  chiType: dataChi.IChiNames;
  value: number;
  cost: ICost;
}

export type IChiAction = IChiBuy;

export function chiBuy(chi: dataChi.IChiNames, value: number, cost: ICost): IChiBuy {
  dataChi.validateChi(chi);
  return { type: constants.CHI_BUY, chiType: chi, value: value, cost: cost };
}
