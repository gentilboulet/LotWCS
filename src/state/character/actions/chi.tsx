import { createAction } from "typesafe-actions";
import { TChiName } from "../../data/chi";
import { ICost } from "../costs";

export const chiBuy = createAction(
  "chi/BUY",
  action => (chi: TChiName, value: number, cost: ICost) => {
    return action({ chi, value, cost });
  }
);
