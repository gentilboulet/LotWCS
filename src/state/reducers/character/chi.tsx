import { Draft, produce } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../../actions/character/chi";
import { TChiState } from "../../models/character/chi";

import { increase } from "../../models/character/chi";

export type IChiAction = ActionType<typeof actions>;

export const chiReducer = produce(
  (draft: Draft<TChiState>, action: IChiAction) => {
    switch (action.type) {
      case getType(actions.chiBuy):
        increase(draft, action.payload.chi, action.payload.value);
        break;
      default:
    }
  },
);
