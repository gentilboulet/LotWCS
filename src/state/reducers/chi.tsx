import { IChiAction } from 'state/actions/chi';
import { IStoreState } from 'state/types';

import * as chi from 'state/chi';
import * as constants from 'state/constants/chi';

// Sub Reducers
import { applyCost } from 'costs/reducer';
import { pushToHistory } from 'state/history';

export function chiReducer(oldState: IStoreState, action: IChiAction): IStoreState {
  switch (action.type) {
    case constants.CHI_BUY:
      return oldState.withMutations(state => {
        applyCost(state, action.cost);

        chi.increaseValue(state, action.chiType, action.value);

        pushToHistory(state, action);
      });
    default:
  }
  return oldState;
}
