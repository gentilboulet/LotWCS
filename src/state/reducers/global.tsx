import { IAction } from 'state/actions/types';
import { IStoreState } from 'state/types';

// Sub Reducers
import * as header from 'constants/header';
import { headerReducer } from 'state/reducers/header';

import * as chi from 'constants/chi';
import { chiReducer } from 'state/reducers/chi';

import * as initial from 'constants/initial';
import { initialStateFactory } from 'state/initial';

import * as kungfus from 'constants/kungfus';
import { kungfuReducer } from 'state/reducers/kungfus';

import * as history from 'constants/history';
import { historyReducer } from 'state/reducers/history';

import * as skills from 'constants/skills';
import { skillsReducer } from 'state/reducers/skills';

import * as loresheets from 'constants/loresheets';
import { loresheetsReducer } from 'state/reducers/loresheets';

import * as virtues from 'constants/virtues';
import { virtuesReducer } from 'state/reducers/virtues';

export function globalReducer(state: IStoreState, action: IAction): IStoreState {
  switch (action.type) {
    case header.HEADER_SET_ARCHETYPE:
    case header.HEADER_SET_CONCEPT:
    case header.HEADER_SET_NAME:
    case header.HEADER_SET_RANK:
      return headerReducer(state, action);
    case chi.CHI_BUY:
      return chiReducer(state, action);
    case initial.INITIAL_STATE:
      return initialStateFactory();
    case kungfus.KUNGFU_OPEN_STYLE:
    case kungfus.KUNGFU_CUSTOM_NAME_FOR_STYLE:
    case kungfus.KUNGFU_CUSTOM_NAME_FOR_TECHNIQUE:
    case kungfus.KUNGFU_BUY_TECHNIQUE:
      return kungfuReducer(state, action);
    case skills.SKILLS_BUY:
      return skillsReducer(state, action);
    case loresheets.LORESHEET_OPEN:
    case loresheets.LORESHEET_BUY_OPTION:
      return loresheetsReducer(state, action);
    case history.HISTORY_DELETE:
      return historyReducer(state, action);
    case virtues.VIRTUE_INCREASE:
      return virtuesReducer(state, action);
    default:
      return state;
  }
}
