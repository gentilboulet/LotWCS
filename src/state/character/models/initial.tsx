/* eslint-disable import/first */
import { createState as createAutomaticsState } from "./automatics";
import { createState as createChiState } from "./chi";
import { createState as createDiscountsState } from "./discounts";
import { createState as createaKungFuState } from "./kungfu";
import { createState as createLoresheetsState } from "./loresheets";
import { createState as createSkillsState } from "./skills";
import { ICharacterState } from "./type";
import { createState as createVirtuesState } from "./virtues";

export function emptyStateFactory(): ICharacterState {
  return {
    archetype: undefined,
    concept: undefined,
    name: undefined,
    rank: undefined,

    destiny: 0,
    entanglement: 0,

    chi: createChiState(),
    kungfu: createaKungFuState(),
    loresheets: createLoresheetsState(),
    skills: createSkillsState(),
    virtues: createVirtuesState(),

    automatics: createAutomaticsState(),
    discounts: createDiscountsState()
  };
}

import { setArchetype, setRank } from "../actions/header";
import { skillSpecialityBuy } from "../actions/skills";
import { playActions } from "../reducers/global";
import { zeroCost } from "./costs";

export function testingStateFactory(): ICharacterState {
  const initial = emptyStateFactory();
  const actions = [
    setRank(2),
    setArchetype("warrior"),
    skillSpecialityBuy("Awareness", "Hear", zeroCost),
    skillSpecialityBuy("Awareness", "Sight", zeroCost)
  ];
  return playActions(initial, actions);
}

export function initialStateFactory(): ICharacterState {
  return emptyStateFactory();
}
