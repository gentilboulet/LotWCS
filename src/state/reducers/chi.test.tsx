import { IChiAction } from 'state/actions/chi';
import { initialStateFactory } from 'state/initial';
import { IStoreState } from 'state/type';

import { chiReducer } from './chi';
import { globalReducer } from './global';

import * as actions from 'state/actions/chi';

const initialState: IStoreState  = initialStateFactory();

describe('Testing chiReducer', () => {
  const noCost = {
    destiny: 0,
    discountIdx: -1,
    discountNewValue: 0,
    entanglement: 0
  };

  it('should receive a CHI_BUY action', () => {
    expect( initialState.chi.fire.value ).toBe(0);;
    const action = actions.chiBuy('fire', 13, noCost);
    const state = chiReducer(initialState, action);
    expect ( state.chi.fire.value ).toBe(13);
    expect( globalReducer(initialState, action) ).toMatchObject(state);
  });

  it('should do nothing with a junk action', () => {
    const junk = { type: 'JUNK_ACTION' };
    expect( chiReducer(initialState, junk as IChiAction )).toMatchSnapshot();
  });
});
