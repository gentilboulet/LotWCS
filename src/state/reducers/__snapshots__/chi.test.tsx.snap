// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing chiReducer should do nothing with a junk action 1`] = `
Object {
  "archetype": "warrior",
  "automatics": Array [],
  "chi": Object {
    "corrupt": Object {
      "cultivation": 0,
      "value": 0,
    },
    "earth": Object {
      "cultivation": 0,
      "value": 0,
    },
    "enlightened": Object {
      "cultivation": 0,
      "value": 0,
    },
    "fire": Object {
      "cultivation": 0,
      "value": 0,
    },
    "general": Object {
      "cultivation": 0,
      "value": 10,
    },
    "metal": Object {
      "cultivation": 0,
      "value": 0,
    },
    "water": Object {
      "cultivation": 0,
      "value": 0,
    },
    "wood": Object {
      "cultivation": 0,
      "value": 0,
    },
  },
  "concept": undefined,
  "destiny": 20,
  "discounts": Array [
    Object {
      "skills": Array [
        "Awareness",
        "Confidence",
        "Crafting",
        "Finesse",
        "Hardiness",
        "Inspire",
        "Learning",
        "Medicine",
        "Might",
        "Perform",
        "Politics",
        "Ride",
        "Stealth",
        "Survival",
        "Tactics",
        "Wu Wei",
      ],
      "type": "DISCOUNT_SKILL",
      "value": 20,
    },
    Object {
      "kfType": "KUNGFU_EXTERNAL",
      "type": "DISCOUNT_KUNGFU_STYLE",
      "uids": Array [],
      "value": 10,
    },
    Object {
      "kfType": "KUNGFU_INTERNAL",
      "type": "DISCOUNT_KUNGFU_STYLE",
      "uids": Array [],
      "value": 10,
    },
  ],
  "entanglement": 0,
  "history": Array [
    Object {
      "error": undefined,
      "meta": undefined,
      "payload": undefined,
      "type": "history/INITIAL_STATE",
    },
    Object {
      "error": undefined,
      "meta": undefined,
      "payload": Object {
        "rank": 2,
      },
      "type": "header/SET_RANK",
    },
    Object {
      "error": undefined,
      "meta": undefined,
      "payload": Object {
        "archetype": "warrior",
      },
      "type": "header/SET_ARCHETYPE",
    },
    Object {
      "error": undefined,
      "meta": undefined,
      "payload": Object {
        "cost": Object {
          "canPay": true,
          "destiny": 0,
          "original": 0,
        },
        "skill": "Awareness",
        "speciality": "Hear",
      },
      "type": "skills/BUY_SPECIALITY",
    },
    Object {
      "error": undefined,
      "meta": undefined,
      "payload": Object {
        "cost": Object {
          "canPay": true,
          "destiny": 0,
          "original": 0,
        },
        "skill": "Awareness",
        "speciality": "Sight",
      },
      "type": "skills/BUY_SPECIALITY",
    },
  ],
  "kungfu": Object {
    "KUNGFU_EXTERNAL": Object {},
    "KUNGFU_INTERNAL": Object {},
    "noRestrictionInternal": Array [],
  },
  "loresheets": Object {},
  "name": undefined,
  "rank": 2,
  "skills": Array [
    Object {
      "name": "Awareness",
      "specialities": Array [
        "Hear",
        "Sight",
      ],
      "value": 0,
    },
    Object {
      "name": "Confidence",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Crafting",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Finesse",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Hardiness",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Inspire",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Learning",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Medicine",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Might",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Perform",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Politics",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Ride",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Stealth",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Survival",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Tactics",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Wu Wei",
      "specialities": Array [],
      "value": 0,
    },
  ],
  "virtues": Array [
    Object {
      "name": "Honor",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Benevolence",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Righteousness",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Loyalty",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Force",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Revenge",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Individualism",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Obsession",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Ruthlessness",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Ferocity",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
  ],
}
`;
